src/UnliftIO/Retry.hs

```diff
@@ -0,0 +1,268 @@
+{-# LANGUAGE RankNTypes            #-}
+
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  UnliftIO.Retry
+-- Copyright   :  Ozgun Ataman <ozgun.ataman@soostone.com>
+-- License     :  BSD3
+--
+-- Maintainer  :  Patrick Brisbin <pbrisbin@gmail.com>
+-- Stability   :  provisional
+--
+-- Unlifted "Control.Retry".
+--
+-- @since 0.9.3.0
+----------------------------------------------------------------------------
+
+
+module UnliftIO.Retry
+    (
+      -- * Types and Operations
+      RetryPolicyM (..)
+    , RetryPolicy
+    , retryPolicy
+    , retryPolicyDefault
+    , natTransformRetryPolicy
+    , RetryAction (..)
+    , toRetryAction
+    , RetryStatus (..)
+    , defaultRetryStatus
+    , applyPolicy
+    , applyAndDelay
+
+
+    -- ** Lenses for 'RetryStatus'
+    , rsIterNumberL
+    , rsCumulativeDelayL
+    , rsPreviousDelayL
+
+    -- * Applying Retry Policies
+    , retrying
+    , retryingDynamic
+    , recovering
+    , recoveringDynamic
+    , stepping
+    , recoverAll
+    , skipAsyncExceptions
+    , logRetries
+    , defaultLogMsg
+    , retryOnError
+    -- ** Resumable variants
+    , resumeRetrying
+    , resumeRetryingDynamic
+    , resumeRecovering
+    , resumeRecoveringDynamic
+    , resumeRecoverAll
+
+    -- * Retry Policies
+    , constantDelay
+    , exponentialBackoff
+    , fullJitterBackoff
+    , fibonacciBackoff
+    , limitRetries
+
+    -- * Policy Transformers
+    , limitRetriesByDelay
+    , limitRetriesByCumulativeDelay
+    , capDelay
+
+    -- * Development Helpers
+    , simulatePolicy
+    , simulatePolicyPP
+    ) where
+
+-------------------------------------------------------------------------------
+import           Control.Retry hiding
+    ( recoverAll
+    , recovering
+    , recoveringDynamic
+    , resumeRecovering
+    , resumeRecoveringDynamic
+    , resumeRecoverAll
+    , stepping
+    )
+import qualified Control.Retry as Retry
+import           Control.Monad.Catch (Handler(..))
```

Comment:
When we did this in the work App, it wasn't a big deal to use `exceptions` `Handler` still, but I think it'd be morally more correct to use `UnliftIO.Exception`'s `Handler`. Doing so should be trivial (localized to `transHandler` below), but it means bringing in `unliftio` now too, (not just `unliftio-core`).

Thoughts?

Reply:
If you were to propagate the `Typeable` constraint when using lens' `handler`, it would all still work, right? As long as you're not totally blocked from using this module, I think it's kind of hard to justify dragging in the extra `unliftio` dependencies.

Reply:
Yeah youou could, but I specifically didn't want to. It was both annoying to do (deep call stacks) and possibly confusing to others on the team. I'm used to `Typeable` being used to, like, print a constructor name or something, so it's odd to see on an `m` like that.

In any event, we stamped it out differently,

```hs
whatever ::  MonadIO m => Handler m X
whatever = transHandler liftIO $ handler _Blah $ ...
```

Reply:
So I guess we leave `UnliftIO.Retry` using `Control.Monad.Catch.Handler`?

Reply:
If all that sounds good to you this seems like it'd be ready for release?

Reply:
Yup, merge away!